openhab/items.py

```diff
@@ -76,6 +89,10 @@ def state(self) -> typing.Any:
   def state(self, value: typing.Any):
     self.update(value)
 
+  @property
+  def members(self):
```

Comment:
Please add a docstring

Reply:
> Have you tested this with nested groups, does this work as expected ?

I've done now :) But it seems to be working. For items as follows:
```
Group FireAlarm

Group:Switch:OR(ON,OFF) FireAlarm_Failure  <error>  (FireAlarm)
Switch B_SmokeSensor_AlarmSystem  <smokedetector> (FireAlarm_Failure) 
Switch B_SmokeSensor_AlarmTamper  <tools> (FireAlarm_Failure)

Group:Switch:OR(ON,OFF) FireAlarm_Fire <fire> (FireAlarm)
Switch B_SmokeSensor_AlarmSmoke  <smoke> (FireAlarm_Fire) 
Switch B_SmokeSensor_AlarmHeat <temperature_hot> (FireAlarm_Fire) 
Number B_SmokeSensor_TimeOffset "Clock Time Offset" (FireAlarm) 
```
(i've truncated above some irrelevant labels and channels declarations)
I'm getting:
```
Python 3.7.1 (v3.7.1:260ec2c36a, Oct 20 2018, 14:57:15) [MSC v.1915 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import openhab
>>>
>>> base_url = 'http://localhost:8080/rest'
>>> openhab = openhab.OpenHAB(base_url)
>>> group_with_groups = openhab.get_item("FireAlarm")
>>> print(group_with_groups)
<None - FireAlarm : None>

>>> print(group_with_groups.members)
{'FireAlarm_Failure': <openhab.items.SwitchItem object at 0x000002D5FF65D160>, 'FireAlarm_Fire': <openhab.items.SwitchItem object at 0x000002D5FF65D128>, 'B_SmokeSensor_TimeOffset': <openhab.items.NumberItem object at 0x000002D5FF65D0B8>}

>>> print(group_with_groups.members['FireAlarm_Failure'].members)
{'B_SmokeSensor_AlarmSystem': <openhab.items.SwitchItem object at 0x000002D5FF65DA90>, 'B_SmokeSensor_AlarmTamper': <openhab.items.SwitchItem object at 0x000002D5FF65DA20>}
```
> I think it would be a good idea to handle the case where a group contains different item types. This seems to be valid per documentation, but the results are said to be unexpected and I guess we should not allow this here (it would break anyway, e.g. switch + number item). I would raise an exception in that case I guess.

Type of the group is defined here by openhab rest API, so it should match declaration on the group item type (and allowed operations). Type of the members of the group is also based on the type of particular member of the group as returned by the API as well. 
I do not see the case where it will break, as types on python will match those in openhab.
If you have in mind some combination of groups and items, where there might be some issue, then let me know. I'm happy to check this and fix.

Reply:
>Please add a docstring

Added in latest commit.

Reply:
> > Have you tested this with nested groups, does this work as expected ?
> 
> I've done now :) But it seems to be working. For items as follows:
> 
> ```
> Group FireAlarm
> 
> Group:Switch:OR(ON,OFF) FireAlarm_Failure  <error>  (FireAlarm)
> Switch B_SmokeSensor_AlarmSystem  <smokedetector> (FireAlarm_Failure) 
> Switch B_SmokeSensor_AlarmTamper  <tools> (FireAlarm_Failure)
> 
> Group:Switch:OR(ON,OFF) FireAlarm_Fire <fire> (FireAlarm)
> Switch B_SmokeSensor_AlarmSmoke  <smoke> (FireAlarm_Fire) 
> Switch B_SmokeSensor_AlarmHeat <temperature_hot> (FireAlarm_Fire) 
> Number B_SmokeSensor_TimeOffset "Clock Time Offset" (FireAlarm) 
> ```
> 
> (i've truncated above some irrelevant labels and channels declarations)
> I'm getting:
> 
> ```
> Python 3.7.1 (v3.7.1:260ec2c36a, Oct 20 2018, 14:57:15) [MSC v.1915 64 bit (AMD64)] on win32
> Type "help", "copyright", "credits" or "license" for more information.
> >>> import openhab
> >>>
> >>> base_url = 'http://localhost:8080/rest'
> >>> openhab = openhab.OpenHAB(base_url)
> >>> group_with_groups = openhab.get_item("FireAlarm")
> >>> print(group_with_groups)
> <None - FireAlarm : None>
> 
> >>> print(group_with_groups.members)
> {'FireAlarm_Failure': <openhab.items.SwitchItem object at 0x000002D5FF65D160>, 'FireAlarm_Fire': <openhab.items.SwitchItem object at 0x000002D5FF65D128>, 'B_SmokeSensor_TimeOffset': <openhab.items.NumberItem object at 0x000002D5FF65D0B8>}
> 
> >>> print(group_with_groups.members['FireAlarm_Failure'].members)
> {'B_SmokeSensor_AlarmSystem': <openhab.items.SwitchItem object at 0x000002D5FF65DA90>, 'B_SmokeSensor_AlarmTamper': <openhab.items.SwitchItem object at 0x000002D5FF65DA20>}
> ```

Great! Looks good


> > I think it would be a good idea to handle the case where a group contains different item types. This seems to be valid per documentation, but the results are said to be unexpected and I guess we should not allow this here (it would break anyway, e.g. switch + number item). I would raise an exception in that case I guess.
> 
> Type of the group is defined here by openhab rest API, so it should match declaration on the group item type (and allowed operations). Type of the members of the group is also based on the type of particular member of the group as returned by the API as well.
> I do not see the case where it will break, as types on python will match those in openhab.
> If you have in mind some combination of groups and items, where there might be some issue, then let me know. I'm happy to check this and fix.

I see, but considering your example 
```
lights_group.on()
```
what would happen if the groups contains a mix of Switch and Number items ?
What does "json_data["groupType"]" look like ?

Reply:
### Short version:
`json_data["groupType"]` will return type of group declared in openhab. So i think it's enough to assume that commands such type can be send to such group.
Allowed mixing item types by openhab i believe is not client side worries. This depends on openhab internal handling and client should not IMHO limit this more than openhab itself and allows in rules.

### Longer version:
For example code in openhab:

- debug.items:
```
Group:Switch:OR(ON,OFF) SwitchGroup
Switch GroupElement_Switch (SwitchGroup)
Number GroupElement_Number (SwitchGroup)
```
- debug.rules
```
rule "Group test"
when
    System started
then
    SwitchGroup.sendCommand(ON)
end
```

> what would happen if the groups contains a mix of Switch and Number items ?

Result from rules triggered is:
```
15:39:56.278 [INFO ] [smarthome.event.ItemCommandEvent     ] - Item 'SwitchGroup' received command ON
15:39:56.279 [INFO ] [smarthome.event.ItemCommandEvent     ] - Item 'GroupElement_Switch' received command ON
15:39:56.280 [INFO ] [smarthome.event.ItemCommandEvent     ] - Item 'GroupElement_Number' received command ON
15:39:56.282 [INFO ] [smarthome.event.ItemStateChangedEvent] - GroupElement_Switch changed from NULL to ON
```
> What does "json_data["groupType"]" look like ?

`json_data["groupType"]` match type in `Group:Switch:OR(ON,OFF) SwitchGroup`
so result will match behavior with rules:
- request:
`curl -X POST --header "Content-Type: text/plain" --header "Accept: application/json" -d "OFF" "http://localhost:8080/rest/items/SwitchGroup"`
- result:
```
15:53:50.897 [INFO ] [smarthome.event.ItemCommandEvent     ] - Item 'SwitchGroup' received command OFF
15:53:50.901 [INFO ] [smarthome.event.ItemCommandEvent     ] - Item 'GroupElement_Switch' received command OFF
15:53:50.904 [INFO ] [smarthome.event.ItemCommandEvent     ] - Item 'GroupElement_Number' received command OFF
15:53:50.911 [INFO ] [smarthome.event.ItemStateChangedEvent] - GroupElement_Switch changed from ON to OFF
```

Reply:
Thanks for your elaborate answer on this. I agree with you that the lib should not be more restrictive than openhab itself. It should feels wrong :-)

I'll go ahead and merge your changes. Thanks again!